In [ ]:
# Imports and pip installations (if needed)
from sklearn.datasets import load_iris
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
import numpy as np
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [ ]:
# Load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)


iris = load_iris()
feature_name = iris['feature_names']
df = pd.DataFrame(iris['data'], columns = feature_name)
df.head(15)



sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                 5.1               3.5                1.4               0.2
1                 4.9               3.0                1.4               0.2
2                 4.7               3.2                1.3               0.2
3                 4.6               3.1                1.5               0.2
4                 5.0               3.6                1.4               0.2
5                 5.4               3.9                1.7               0.4
6                 4.6               3.4                1.4               0.3
7                 5.0               3.4                1.5               0.2
8                 4.4               2.9                1.4               0.2
9                 4.9               3.1                1.5               0.1
10                5.4               3.7                1.5               0.2
11                4.8               3.4                1.6               0.2
12                4.8               3.0                1.4               0.1
13                4.3               3.0                1.1               0.1
14                5.8               4.0                1.2               0.2

In [ ]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data is showing a dataset of Iris Flowers and it is presenting the different attributes each flower has. If we look at the dataframe table we can see the features are sepal length and width, and petal length and width. The data also contains a total of 150 datapoints. We can also see the range of values for each column with the df.describe() function.

# Part 2: Split the dataset into train and test

In [ ]:
# Take the dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
target = 'species'
df[target] = iris['target']

X = df[feature_name]
Y = df[target]

X_train, X_test, y_train, y_test = train_test_split(X, Y,test_size=0.1)

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
# # ii. For a sample datapoint, predict the probabilities for each possible class
# iii. Report on the score for Logistic regression model, what does the score measure?
# iv. Extract the coefficents and intercepts for the boundary line(s)

regression = LogisticRegression()
regression.fit(X_train, y_train)

sample = np.array([[3, 3, 3, 3]])
print("Predicted Probabilities: ", regression.predict_proba(sample)[0])

scoreReg = regression.score(X_test, y_test)
print("Score: ", scoreReg)

coefficent = regression.coef_
intercept = regression.intercept_
print("Coefficients: ", coefficent)
print("Intercept: ", intercept)

Predicted Probabilities:  [0.61284794 0.15268909 0.23446298]
Score:  1.0
Coefficients:  [[-0.44666665  0.91105163 -2.4334074  -1.04125939]
 [ 0.52280697 -0.317622   -0.20334196 -0.90588931]
 [-0.07614032 -0.59342963  2.63674935  1.94714869]]
Intercept:  [  9.81435616   2.10593653 -11.92029269]


E:\Users\Sakib Amin\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


The score measures the accuracy of the testing dataset. The program calculate accuracy by divding the number of correctly predicted with the total number of datapoints in the set. Since we got a score of 1 it means the model was able to correctly predict all the datapoints. 

# Part 4: Support Vector Machine

In [ ]:
# i. Use sklearn to train a Support Vector Classifier on the training set
# ii. For a sample datapoint, predict the probabilities for each possible class
# iii. Report on the score for the SVM, what does the score measure?
svc = SVC(probability=True)
svc.fit(X_train, y_train)

sample = np.array([[3,3,3,3]])
print("Predicted Probabilities:", svc.predict_proba(sample)[0])

scoreSVC = svc.score(X_test, y_test)
print("Score: ", scoreSVC)

Predicted Probabilities: [0.17798648 0.49153263 0.3304809 ]
Score:  1.0


Similar to the score of Part 3 this score does the same thing as it is checking accurarcy of the testing dataset. We got a score of 1 in this model meaning we have 100% accurary in predicting all the datapoints

# Part 5: Neural Network

In [ ]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
# ii. For a sample datapoint, predict the probabilities for each possible class
# iii. Report on the score for the Neural Network, what does the score measure?
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

MLP = MLPClassifier()
MLP.fit(X_train, y_train)

sample = np.array([[3,3,3,3]])
print("Predicted Probabilities:", MLP.predict_proba(sample)[0])

scoreMLP = MLP.score(X_test, y_test)
print("Score: ", scoreMLP)

MLP2 = MLPClassifier(
            activation='logistic',
            solver='lbfgs',
            alpha=0.1,
            tol=0.001
)
MLP2.fit(X_train, y_train)
scoreMLP2 = MLP2.score(X_test, y_test)
print("Modified Score: ", scoreMLP2)

Predicted Probabilities: [0.01524949 0.0302322  0.95451831]
Score:  1.0
Modified Score:  1.0


E:\Users\Sakib Amin\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


The score is measuring the accuracy of the model. For both the score and the modiefied score I got a score of 1 meaning the model can predict the correct answer 100% of the time

# Part 6: K-Nearest Neighbors

In [ ]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
# ii. For a sample datapoint, predict the probabilities for each possible class

knn = KNeighborsClassifier()
knn.fit(X_train, y_train)

print("Predicted Probabilities:", knn.predict_proba(sample)[0])

Scoreknn = knn.score(X_test, y_test)
print("Score: ", Scoreknn)

Predicted Probabilities: [0.  0.8 0.2]
Score:  1.0


Once again the score I achieved was 1 meaning that the model was able to predict the correct answer 100% of the time

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

After writing the code and running all the models, I have recieved a 100% accuracy on all the models. This was surprising to me because I expected some variance between the different models but I believe this is probably due to the small dataset that we have of only 150 Iris flower. Maybe with a bigger dataset it will cause more variation in the dataset. 